### PR TITLE
[ENG-6036] Implement DataverseStorageAddonImp

### PR DESCRIPTION
**Pull request by opaduchak**

### CODE REVIEW

File: addon_imps/storage/dataverse.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    @staticmethod
+    def make_headers(raw_headers: dict):
+        header_value = raw_headers.get("Authorization")
+        return {"X-Dataverse-key": header_value.split(" ")[-1]}
+
+    def __post_init__(self):
+        self.network.add_header_transformation_callback(self.make_headers)
+
+    async def get_external_account_id(self, _: dict[str, str]) -> str:
+        return ""
```
Comment (aaxelb) [inline addon_imps/storage/dataverse.py:39]: nit: this should be unnecessary, it's already implemented by the base class

**Review by aaxelb — COMMENTED** [review]
comment on the "header transformation", but haven't dug into dataverse specifics

File: addon_imps/storage/dataverse.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    see https://guides.dataverse.org/en/latest/api/native-api.html
+    """
+
+    @staticmethod
+    def make_headers(raw_headers: dict):
+        header_value = raw_headers.get("Authorization")
+        return {"X-Dataverse-key": header_value.split(" ")[-1]}
+
+    def __post_init__(self):
+        self.network.add_header_transformation_callback(self.make_headers)
```
[the remark this answers is not in the document]
Comment (opaduchak) [reply addon_imps/storage/dataverse.py]: Fixed credentials handling, please look once more and tell me whether this is suffiecient

File: addon_imps/storage/dataverse.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+    async def get_external_account_id(self, _: dict[str, str]) -> str:
+        return ""
+
+    async def list_root_items(self, page_cursor: str = "") -> storage.ItemSampleResult:
+        async with self.network.GET(
+            "mydata/retrieve",
+            query=[
+                ["page", page_cursor],
+                *[("role_ids", role) for role in range(1, 9)],
```
Comment (aaxelb) [inline addon_imps/storage/dataverse.py:38]: dataverse docs say roles are configurable per instance... do we have a sense whether they actually vary in practice, or is it safe to hard-code this?

**Review by aaxelb — APPROVED** [review]
the credentials changes seem a good improvement! i don't know the dataverse api well, but the code looks reasonable 👍

File: addon_imps/storage/dataverse.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaxelb) [reply addon_imps/storage/dataverse.py]: looks great!

File: addon_imps/storage/dataverse.py (hunk 3 above)
Comment (opaduchak) [reply addon_imps/storage/dataverse.py:38]: these are all of the standart roles which are present in all dataverses

File: addon_toolkit/constrained_network/http.py [hunk 4]
```diff
@@ -62,6 +63,10 @@ class HttpRequestor(typing.Protocol):
     @property
     def response_info_cls(self) -> type[HttpResponseInfo]: ...
 
+    def add_header_transformation_callback(
+        self, callback: Callable[[dict], dict]
+    ) -> None: ...
+
```
Comment (aaxelb) [inline addon_toolkit/constrained_network/http.py:69]: this is no longer used, can be removed

**Review by aaxelb — COMMENTED** [review]
just a bit of dead code, otherwise seems good

File: addon_service/common/known_imps.py [hunk 5]
```diff
@@ -87,6 +88,9 @@ class AddonImpNumbers(enum.Enum):
     GOOGLE_DRIVE = 1005
     FIGSHARE = 1007
     ONEDRIVE = 1008
+    DATAVERSE = 1010
+
+
```
Comment (brianjgeiger) [inline addon_service/common/known_imps.py:93]: Looks like this is causing problems with pre-commit; you should probably delete these two lines.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Just one small change to make CI happy.

**Review by brianjgeiger — APPROVED** [review]

**Pull request merged**